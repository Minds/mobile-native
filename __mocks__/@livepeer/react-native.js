// mock @livepeer/react-native
export const LivepeerConfig = {
  setConfig: jest.fn(),
  getConfig: jest.fn(),
  get: jest.fn(),
  set: jest.fn(),
};

export const createReactClient = jest.fn();
export const createLivepeerClient = jest.fn();
export const Player = jest.fn();
