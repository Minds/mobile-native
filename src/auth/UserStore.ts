import { observable, action } from 'mobx';

import channelService from './../channel/ChannelService';
import UserModel from './../channel/UserModel';
import searchBarService from '../topbar/searchbar/SearchBar.service';

/**
 * Login Store
 */
class UserStore {
  @observable me: UserModel;
  @observable emailConfirmMessageDismiss = false;
  @observable searching = false;
  @observable bannerInfoDismiss = false;
  @observable rewards: any = false;

  constructor() {
    this.me = UserModel.create({});
  }

  /**
   * Dissmis Message Banner
   */
  @action
  setDissmisBanner(value: boolean): void {
    this.bannerInfoDismiss = value;
  }

  @action
  setDissmis(value: boolean): void {
    this.emailConfirmMessageDismiss = value;
  }

  @action
  setUser(user: { guid?: string }): void {
    if (!user || !user.guid) {
      return;
    }
    this.me = UserModel.create(user);
  }

  @action
  setRewards(value: boolean) {
    this.me.rewards = value;
  }

  @action
  setWallet(value) {
    this.me.eth_wallet = value;
  }

  @action
  setTosLastUpdate(value) {
    this.me.last_accepted_tos = value;
  }

  hasRewards() {
    return !!(this.me && this.me.rewards);
  }

  hasEthWallet() {
    return !!(this.me && this.me.eth_wallet);
  }

  hasAvatar() {
    return this.me.icontime !== this.me.time_created;
  }

  @action
  clearUser() {
    this.me = UserModel.create({});
  }

  @action
  async load(refresh = false) {
    if (!refresh) {
      this.clearUser();
    }

    let response: any = await channelService.load('me');

    this.setUser(response.channel);

    // Load search history
    searchBarService.init(this.me.guid);

    return this.me;
  }

  isAdmin() {
    return this.me.isAdmin();
  }

  reset() {
    this.clearUser();
  }

  @action
  toggleSearching = () => {
    this.searching = !this.searching;
  };

  /**
   * Call onItemTap
   */
  searchBarItemTap(item: UserModel | string) {
    if (item instanceof UserModel) {
      const user = { user: item.toPlainObject() };
      searchBarService.onItemTap(user);
    } else {
      searchBarService.onItemTap(item);
    }
  }

  /**
   * Clear search history for user
   */
  searchBarClearHistory() {
    searchBarService.clearSearchHistory();
  }

  /**
   * Get user search history
   */
  async getSearchHistory() {
    return await searchBarService.getSearchHistory();
  }

  /**
   * Get suggested Search
   * @param {String} search
   */
  async getSuggestedSearch(search) {
    return await searchBarService.getSuggestedSearch(search);
  }
}

export default UserStore;
