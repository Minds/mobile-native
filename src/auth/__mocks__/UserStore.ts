import { extendObservable } from 'mobx';

const mock = jest.fn().mockImplementation(() => {
  return extendObservable(
    {
      setUser: jest.fn(),
      setRewards: jest.fn(),
      clearUser: jest.fn(),
      load: jest.fn(),
      isAdmin: jest.fn(),
      reset: jest.fn(),
    },
    {
      me: {},
      hasAttachment: false,
    },
  );
});

export default mock;
