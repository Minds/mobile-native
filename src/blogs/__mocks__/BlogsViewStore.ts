import { extendObservable } from 'mobx';

const mock = jest.fn().mockImplementation(() => {
  return extendObservable(
    {
      loadBlog: jest.fn(),
      setBlog: jest.fn(),
      reset: jest.fn(),
    },
    {
      blog: null,
    },
  );
});

export default mock;
