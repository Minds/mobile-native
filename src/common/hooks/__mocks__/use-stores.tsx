import { useLocalStore } from 'mobx-react';
import createPortraitStore from '../../../portrait/createPortraitStore';
import createSearchResultStore from '../../../topbar/searchbar/createSearchResultStore';
import createWalletStore from '../../../wallet/v2/createWalletStore';
import { createLegacyStores, TLegacyStores } from '../../contexts';
import { StoresType } from '../use-stores';

/**
 * Allows for a function component to consume our global stores
 * This **MUST** be only consumed below <StoresProvider> (which is placed in src/App.tsx)
 */
export const useStores = (): StoresType => {
  const stores = {
    wallet: useLocalStore(createWalletStore),
    searchBar: useLocalStore(createSearchResultStore),
    portrait: useLocalStore(createPortraitStore),
  };
  return stores;
};

/**
 * Allows for a function component to consume our legacy global stores that have been
 * provided via mobx's <Provider>
 */
export const useLegacyStores = (): TLegacyStores => {
  return createLegacyStores();
};
