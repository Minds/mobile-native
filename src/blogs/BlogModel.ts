import { ImageURISource } from 'react-native';
import { decorate, observable } from 'mobx';
import { MINDS_ASSETS_CDN_URI } from '../config/Config';
import api from '../common/services/api.service';
import { LICENSES } from '../common/services/list-options.service';
import ActivityModel from '../newsfeed/ActivityModel';

/**
 * User model
 */
export default class BlogModel extends ActivityModel {
  license: string = '';
  slug: string = '';

  /**
   * Get banner source
   */
  getBannerSource(): ImageURISource {
    const uri =
      this.thumbnail_src &&
      this.thumbnail_src !== 'https://cdn.minds.com/thumbProxy?src=&c=2708'
        ? this.thumbnail_src
        : `${MINDS_ASSETS_CDN_URI}front/dist/assets/logos/placeholder-bulb.jpg`;

    return { uri, headers: api.buildHeaders() };
  }

  /**
   * Get the license text
   */
  getLicenseText(): string {
    const lic = LICENSES.find(license => this.license === license.value);
    if (!lic) {
      return '';
    }
    return lic.text;
  }
}

/**
 * Define model observables
 */
//@ts-ignore
decorate(BlogModel, {
  description: observable,
});
