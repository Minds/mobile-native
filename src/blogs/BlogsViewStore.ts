import { observable, action } from 'mobx';
import blogService from './BlogsService';
import BlogModel from './BlogModel';
import CommentsStore from '../comments/v2/CommentsStore';

/**
 * Blogs View Store
 */
class BlogsViewStore {
  @observable.ref blog: BlogModel | null = null;

  comments: CommentsStore | null = null;

  /**
   * Load blog
   * @param {string} guid
   */
  async loadBlog(guid) {
    const result = await blogService.loadEntity(guid);
    // keep the _list if the entity has one
    if (this.blog) {
      this.blog.update(result.blog);
    } else {
      await this.setBlog(result.blog);
    }
  }

  /**
   * Reset the store
   */
  @action
  reset() {
    this.blog = null;
  }

  /**
   * Set blog
   * @param {object} blog
   */
  @action
  async setBlog(blog) {
    const instance = BlogModel.checkOrCreate(blog);
    this.comments = new CommentsStore(instance);

    // if the blog is paywalled we try to unlock it
    if (instance.paywall) {
      await instance.unlock();
    }
    this.blog = instance;
  }
}

export default BlogsViewStore;
