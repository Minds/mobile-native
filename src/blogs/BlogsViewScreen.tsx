import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { inject, observer } from 'mobx-react';
import React, { Component } from 'react';
import { Alert, Image, SafeAreaView, ScrollView, View } from 'react-native';
import { Icon } from 'react-native-elements';
import { SafeAreaInsetsContext } from 'react-native-safe-area-context';
import assets from '@assets';

import SmallCircleButton from '~/common/components/SmallCircleButton';
import withModalProvider from '~/navigation/withModalProvide';
import Actions from '~/newsfeed/activity/Actions';
import { pushCommentBottomSheet } from '../comments/v2/CommentBottomSheet';
import CenteredLoading from '../common/components/CenteredLoading';
import MText from '../common/components/MText';
import SmartImage from '../common/components/SmartImage';
import { FLAG_VIEW } from '../common/Permissions';
import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import { AppStackParamList } from '../navigation/NavigationTypes';
import OwnerBlock from '../newsfeed/activity/OwnerBlock';
import { ComponentsStyle } from '../styles/Components';
import ThemedStyles from '../styles/ThemedStyles';
import Lock from '../wire/v2/lock/Lock';
import BlogActionSheet from './BlogActionSheet';
import BlogsViewStore from './BlogsViewStore';
import BlogViewHTML from './BlogViewHTML';
import { withErrorBoundaryScreen } from '~/common/components/ErrorBoundaryScreen';

type BlogScreenRouteProp = RouteProp<AppStackParamList, 'BlogView'>;
type BlogScreenNavigationProp = StackNavigationProp<
  AppStackParamList,
  'BlogView'
>;

type PropsType = {
  blogsView: BlogsViewStore;
  route: BlogScreenRouteProp;
  navigation: BlogScreenNavigationProp;
};

/**
 * Blog View Screen
 */
@inject('user')
@observer
class BlogsViewScreen extends Component<PropsType> {
  commentsRef: any;
  blogsView: BlogsViewStore;

  /**
   * Disable navigation bar
   */
  static navigationOptions = {
    header: null,
  };

  state = {
    error: null,
  };

  /**
   * Constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);

    this.blogsView = props.blogsView ?? new BlogsViewStore();
  }

  /**
   * Component did mount
   */
  async componentDidMount() {
    const params = this.props.route.params;
    try {
      if (params.blog) {
        await this.blogsView.setBlog(params.blog);

        if (!params.blog.description) {
          await this.blogsView.loadBlog(params.blog.guid);
        }
      } else {
        this.blogsView.reset();
        let guid;
        if (params.slug) {
          guid = params.slug.substr(params.slug.lastIndexOf('-') + 1);
        } else {
          guid = params.guid;
        }

        await this.blogsView.loadBlog(guid);
      }

      // check permissions
      if (!this.blogsView.blog?.can(FLAG_VIEW, true)) {
        this.props.navigation.goBack();
        return;
      }

      if (this.blogsView.blog) {
        this.blogsView.blog.sendViewed('single');
      }
    } catch (error: any) {
      logService.exception(error);
      Alert.alert(
        'Error',
        error.message || i18n.t('blogs.errorLoading'),
        [{ text: i18n.t('ok'), onPress: () => this.props.navigation.goBack() }],
        { cancelable: false },
      );
    }
  }

  /**
   * On component will unmount
   */
  componentWillUnmount() {
    this.blogsView.reset();
  }

  getToolbar() {
    const blog = this.blogsView.blog;
    if (!blog) {
      return null;
    }
    const theme = ThemedStyles.style;

    return (
      <SafeAreaView style={theme.bgPrimaryBackground}>
        <Actions
          onPressComment={() => {
            if (this.blogsView?.comments) {
              pushCommentBottomSheet({
                commentsStore: this.blogsView.comments,
              });
            }
          }}
          entity={blog}
        />
      </SafeAreaView>
    );
  }

  /**
   * Render blog
   */
  getBody() {
    const blog = this.blogsView.blog;
    if (!blog) {
      return null;
    }
    const theme = ThemedStyles.style;
    const image = blog.getBannerSource();

    return (
      <ScrollView stickyHeaderIndices={[1]}>
        <>
          <SmartImage source={image} contentFit="cover" style={styles.image} />
          <MText style={styles.title}>{blog.title}</MText>
          <SafeAreaView style={styles.header}>
            <SmallCircleButton
              name="chevron-left"
              raised
              size={20}
              style={theme.colorIcon}
              onPress={this.props.navigation.goBack}
              iconStyle={styles.iconStyle}
            />
          </SafeAreaView>
        </>

        <SafeAreaInsetsContext.Consumer>
          {insets => (
            <View
              style={[styles.ownerBlockContainer, { paddingTop: insets?.top }]}>
              <OwnerBlock
                entity={blog}
                navigation={this.props.navigation}
                rightToolbar={
                  <View style={styles.actionSheet}>
                    <BlogActionSheet
                      entity={blog}
                      navigation={this.props.navigation}
                    />
                  </View>
                }>
                <MText style={[styles.timestamp, theme.colorSecondaryText]}>
                  {i18n.date(parseInt(blog.time_created, 10) * 1000)}
                </MText>
              </OwnerBlock>
            </View>
          )}
        </SafeAreaInsetsContext.Consumer>

        <View>
          <View style={styles.description}>
            {blog.description ? (
              <BlogViewHTML html={blog.description} />
            ) : blog.paywall ? (
              // FIXME: Lock text is white on white and overlaps with license
              <Lock entity={blog} navigation={this.props.navigation} />
            ) : (
              <CenteredLoading />
            )}
          </View>
          {!blog.paywall && (
            <View style={styles.moreInformation}>
              {Boolean(blog.getLicenseText()) && (
                <Icon style={theme.colorIcon} size={18} name="public" />
              )}
              <MText
                style={[
                  theme.fontXS,
                  theme.paddingLeft,
                  theme.colorSecondaryText,
                  theme.paddingRight2x,
                ]}>
                {blog.getLicenseText()}
              </MText>
            </View>
          )}
        </View>
      </ScrollView>
    );
  }

  /**
   * Show an error message
   */
  showError() {
    Alert.alert(
      i18n.t('sorry'),
      i18n.t('errorMessage') + '\n' + i18n.t('activity.tryAgain'),
      [
        {
          text: i18n.t('ok'),
          onPress: () => {},
        },
      ],
      { cancelable: false },
    );
  }

  /**
   * Render
   */
  render() {
    const theme = ThemedStyles.style;

    // TODO: add loading state
    if (!this.blogsView.blog) {
      return <CenteredLoading />;
    } else {
      // force observe on description
      this.blogsView.blog.description;
    }

    // check async update of permissions
    if (!this.blogsView.blog.can(FLAG_VIEW, true)) {
      this.props.navigation.goBack();
      return null;
    }

    return (
      <View style={styles.container}>
        {!this.state.error ? (
          <>
            {this.getBody()}
            {this.getToolbar()}
          </>
        ) : (
          <View style={theme.flexColumnCentered}>
            <Image
              resizeMode="contain"
              style={ComponentsStyle.logo}
              source={assets.LOGO_SQUARED}
            />
            <MText style={[theme.fontL, theme.colorAlert]}>
              {i18n.t('blogs.error')}
            </MText>
            <MText style={[theme.fontM]}>{i18n.t('activity.tryAgain')}</MText>
          </View>
        )}
      </View>
    );
  }
}

export default withErrorBoundaryScreen(BlogsViewScreen, 'BlogsViewScreen');

export const withModal = withModalProvider(
  withErrorBoundaryScreen(BlogsViewScreen, 'BlogsViewScreen'),
);

/**
 * Styles
 */
const styles = ThemedStyles.create({
  actionSheet: {
    paddingLeft: 5,
  },
  header: {
    position: 'absolute',
    left: 0,
    top: 0,
    width: '100%',
  },
  title: {
    paddingTop: 12,
    paddingBottom: 8,
    paddingLeft: 15,
    paddingRight: 15,
    fontSize: 22,
    // fontWeight: '800',
    fontFamily: 'Roboto_900Black', // workaround android ignoring >= 800
  },
  ownerBlockContainer: ['bgPrimaryBackground'],
  description: {
    paddingLeft: 15,
    paddingRight: 15,
    paddingBottom: 15,
  },
  image: {
    height: 200,
  },
  timestamp: {
    fontSize: 11,
    color: '#888',
  },
  moreInformation: {
    padding: 12,
    flexDirection: 'row',
  },
  container: ['flexContainer', 'bgPrimaryBackground'],
  iconStyle: { fontSize: 28 },
});
