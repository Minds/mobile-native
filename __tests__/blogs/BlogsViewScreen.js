/* eslint-disable jest/no-disabled-tests */
import 'react-native';
import React from 'react';
import BlogsViewScreen from '../../src/blogs/BlogsViewScreen';
import BlogsViewStore from '../../src/blogs/BlogsViewStore';
import blogsFaker from '../../__mocks__/fake/blogs/BlogFactory';
import BlogModel from '../../src/blogs/BlogModel';
import UserStore from '../../src/auth/UserStore';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import NavigationService from '../../src/navigation/NavigationService';
import { StoresProvider } from '../../src/common/hooks/use-stores';

jest.mock('../../src/navigation/NavigationService');
jest.mock('../../src/blogs/BlogsViewStore');
jest.mock('../../src/auth/UserStore');
jest.mock('../../src/blogs/BlogViewHTML', () => 'BlogViewHTML');
jest.mock(
  '../../src/comments/v2/CommentBottomSheet',
  () => 'CommentBottomSheet',
);
jest.mock(
  '../../src/newsfeed/activity/actions/RemindAction',
  () => 'RemindAction',
);
jest.mock(
  '../../src/newsfeed/activity/actions/ThumbAction',
  () => 'ThumbAction',
);
jest.mock(
  '../../src/newsfeed/activity/actions/CommentsAction',
  () => 'CommentsAction',
);

jest.mock(
  'react-native-safe-area-context',
  () => require('react-native-safe-area-context/jest/mock').default,
);

jest.mock('../../src/common/services/analytics.service');

/**
 * Tests
 */
describe('blog view screen component', () => {
  let store,
    user,
    route,
    navigation = {
      goBack: jest.fn(),
    };

  beforeEach(() => {
    NavigationService.getCurrentState.mockClear();
    NavigationService.getCurrentState.mockReturnValue({});
    store = new BlogsViewStore();
    user = new UserStore();
    user.me = {
      guid: 'guidguid',
    };
  });

  it.skip('should renders correctly', () => {
    store.blog = BlogModel.create(blogsFaker('1'));

    route = { params: { guid: 1 } };

    const component = renderer
      .create(
        <StoresProvider>
          <BlogsViewScreen.wrappedComponent
            blogsView={store}
            navigation={navigation}
            route={route}
            user={user}
          />
        </StoresProvider>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it.skip('should load the blog by guid', () => {
    store.blog = BlogModel.create(blogsFaker('1'));

    route = { params: { guid: 1 } };

    const component = renderer.create(
      <StoresProvider>
        <BlogsViewScreen.wrappedComponent
          blogsView={store}
          navigation={navigation}
          route={route}
          user={user}
        />
      </StoresProvider>,
    );

    expect(store.loadBlog).toBeCalledWith(1);
  });

  it.skip('should set the blog from params', () => {
    store.blog = BlogModel.create(blogsFaker('1'));

    route = { params: { blog: store.blog } };

    const component = renderer.create(
      <StoresProvider>
        <BlogsViewScreen.wrappedComponent
          blogsView={store}
          navigation={navigation}
          route={route}
          user={user}
        />
      </StoresProvider>,
    );

    expect(store.setBlog).toBeCalledWith(store.blog);
  });
});
