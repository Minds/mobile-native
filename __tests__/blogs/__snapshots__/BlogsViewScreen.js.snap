// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blog view screen component should renders correctly 1`] = `null`;
